### PR TITLE
Correct `$(IntermediateDepsFilePath)` logic

### PR DESCRIPTION
**Pull request by dougbu**
- restore previous behaviour of this target in the non-single-file case
  - ensure `$(PublishDepsFilePath)` exists after `GeneratePublishDependencyFile` target runs

### CODE REVIEW

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets
Comment (swaroop-sridhar) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets:994]: If the deps.json file is directly written to the publish directory, it should be removed from ResolvedFileToPublish [here](https://github.com/dotnet/sdk/blob/c09668ef26a1c8917144e897b2989524cc332d47/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets#L1035) in the non-single-file case.

Comment (swaroop-sridhar) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets:994]: However, I'm not sure its a good idea to write to the publish directory -- before `CopyFilesToPublishDirectory` has run -- because something may be introduced into `ResolvedFileToPublish` that overwrites this file. Isn't it? 

CC: @dsplaisted

**Review by swaroop-sridhar — COMMENTED** [review]
@dougbu, in this case, I think it is better for [InjectRequestHandlerOnPublish](https://github.com/dotnet/aspnetcore/blob/ba4dc9f347658975e6c9aa2a93b22fef3193681e/src/Servers/IIS/build/testsite.props#L62-L74) to run after CopyFilesToPublishDirectory which will copy the deps file from `IntermediateDepsFilePath` to `PublishDepsFilePath`.

If this cannot change due to legacy reasons, then please apply the change I've suggested here.

**Comment by dougbu** [comment]
> If this cannot change due to legacy reasons

Won't external developers get tripped up by this breaking change too? For the non-single-file case, running after `GeneratePublishDependencyFile` and using `$(PublishDepsFilePath)` worked perfectly. But, I defer to you and @dotnet/nsboard-platform. Thoughts?

**Comment by dougbu** [comment]
Oops, whoever I accidentally tagged instead of @dsplaisted, my apologies.

**Comment by dougbu** [comment]
Hmm, what's up w/ the pipeline failure? Did a credential just expire?

**Comment by dougbu** [comment]
Oh good, credential problem went away.

**Comment by dougbu** [comment]
:up::date: to take @swaroop-sridhar's suggestion in case we decide to fix the latest break. Otherwise I'll use `$(IntermediateDepsFilePath)` in dotnet/aspnetcore#21017. It doesn't feel right to leave an incorrect .deps file in the obj/ directory and that would happen if I moved our fixup after the `CopyFilesToPublishDirectory` target. (Must say it also doesn't feel right to leave a .deps file in the obj/ directory at all in the non-single-file case.)

**Review by swaroop-sridhar — APPROVED** [review]

**Comment by dougbu** [comment]
@swaroop-sridhar should I merge this change after PR validation completes or wait for word from @dsplaisted on his preferences?

**Comment by dsplaisted** [comment]
So we're now hitting a different break than we fixed in #11462, right?

The [ASP.NET targets](https://github.com/dotnet/aspnetcore/blob/ba4dc9f347658975e6c9aa2a93b22fef3193681e/src/Servers/IIS/build/testsite.props#L62-L74):

- Run after `GenerateBuildDependencyFile` and `GeneratePublishDependencyFile`
- Expect `PublishDepsFilePath` to be the path to a deps file that they can rewrite

I think that means to minimize breakage, we would have to:

- Have `PublishDepsFilePath` point to the path in the output folder
- Not use `ResolvedFileToPublish` to copy to the output folder, because then targets such as ASP.NET may run before it's copied there

It's not feasible for us to make that true for single-file apps, so existing targets could still get broken in that case.

Is that all correct?  Is that what these changes are trying to achieve?

Is it possible to update the ASP.NET targets to work with the current logic?

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets
Comment (swaroop-sridhar) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets:989]: Can you please update the comments here to say:
In the single-file case, we write deps file to intermediate dir, and it is bundled into the single-file app.
In the non-single-file case, we write deps file directly in the publish directory due to legacy reasons.

**Comment by swaroop-sridhar** [comment]
> So we're now hitting a different break than we fixed in #11462, right?
> 
> The [ASP.NET targets](https://github.com/dotnet/aspnetcore/blob/ba4dc9f347658975e6c9aa2a93b22fef3193681e/src/Servers/IIS/build/testsite.props#L62-L74):
> 
> * Run after `GenerateBuildDependencyFile` and `GeneratePublishDependencyFile`
> * Expect `PublishDepsFilePath` to be the path to a deps file that they can rewrite
> 
> I think that means to minimize breakage, we would have to:
> 
> * Have `PublishDepsFilePath` point to the path in the output folder
> * Not use `ResolvedFileToPublish` to copy to the output folder, because then targets such as ASP.NET may run before it's copied there
> 
> It's not feasible for us to make that true for single-file apps, so existing targets could still get broken in that case.
> 
> Is that all correct? Is that what these changes are trying to achieve?
> 
> Is it possible to update the ASP.NET targets to work with the current logic?

Yes this is accurate. 

As @dougbu  has mentioned, it is possible to fix asp.net targets by using `IntermediateDepsFilePath` instead of `PublishDepsFilePath`. 
Or, by using `PublishDepsFilePath` after `CopyResolvedFilesToPublishDirectory` has run. 

This change is trying to preserve legacy behavior for asp.net and other customers who may use these targets/properties not publishing for non-single-file.

**Comment by dougbu** [comment]
> Can you please update the comments here to say&hellip;

I'll do this if we decide to take this fix. It sounds like @dsplaisted is unconvinced😸

**Comment by dsplaisted** [comment]
I'm unconvinced in the sense that I'm not really sure what we should do.

I think @marcpopMSFT and/or @KathleenDollard were looking at having a process where we at least document breaking changes like this.

So, I think we should fix ASP.NET to work with the current logic to unblock ourselves, and have a discussion about what we should do with this break.

**Comment by dougbu** [comment]
:ok: I'll leave this open rather than fixing the comment or the busted tests (`PublishDepsFilePathIsSetAsExpectedForNormalApps` and `It_cleans_before_trimmed_single_file_publish`) immediately. We can use this if we decide to make a change and close it otherwise.

**Comment by dougbu** [comment]
Shoot. I wasn't thinking things through. Using `$(IntermediateDepsFilePath)` won't work because we also want the path to help us find the native DLL to copy. I'll have to switch to running our target after `CopyResolvedFilesToPublishDirectory` even though I'd rather not leave an incorrect (and, to me, useless) .deps file in the obj/ folder.

**Comment by swaroop-sridhar** [comment]
> Shoot. I wasn't thinking things through. Using `$(IntermediateDepsFilePath)` won't work because we also want the path to help us find the native DLL to copy. I'll have to switch to running our target after `CopyResolvedFilesToPublishDirectory` even though I'd rather not leave an incorrect (and, to me, useless) .deps file in the obj/ folder.

I think leaving behind an "incorrect" .deps file in obj folderis not a major concern. It is actually typical of post-processing tools to modify outputs between obj and bin directories. For example, R2R takes files in `obj` directory, modifies them (elsewhere) and copies the modified output to `bin` directory. So, the `bin` and `obj` directories will not always match.

**Comment by dougbu** [comment]
> I think leaving behind an "incorrect" .deps file in obj folderis not a major concern.

I understand that reasoning though the breaking change part of it and the utility of the new obj/ file still feels **Just Wrong:tm:** 😉

**Comment by dougbu** [comment]
/fyi the `CopyResolvedFilesToPublishDirectory` target doesn't exist and is a typo in a few comments above. Believe `CopyFilesToPublishDirectory` is correct.

**Comment by dougbu** [comment]
@dsplaisted @swaroop-sridhar @marcpopMSFT @KathleenDollard have you had a chance to discuss whether to revert this breaking change or documented it?

**Comment by swaroop-sridhar** [comment]
I recommend that we close this PR, and document the change in `PublishDepsFilePath` semantics.

**Comment by dougbu** [comment]
@swaroop-sridhar nobody goes looking in the regular docs for notifications of breaking changes. Suggest you need an announcement too.

**Pull request closed without merging**
